### PR TITLE
Fix source index allocation in CompilerStack

### LINKED ISSUE #2142: AST inconsistency between gcc and clang
clang:
```
{"children":[{"attributes":{"fullyImplemented":true,"isLibrary":false,"linearizedBaseContracts":[1],"name":"A"},"children":[],"id":1,"name":"ContractDefinition","src":"0:14:0"}],"name":"SourceUnit"}
```

gcc:
```
{"children":[{"attributes":{"fullyImplemented":true,"isLibrary":false,"linearizedBaseContracts":[1],"name":"A"},"children":[],"id":1,"name":"ContractDefinition","src":"0:14:1"}],"name":"SourceUnit"}
```

Note the `src` field differs, `0:14:0` vs `0:14:1`.

### PR DESCRIPTION
**Pull request by axic**
Fixes #2142.

### CODE REVIEW

**Review by chfast — APPROVED** [review]
This fix is ok, but I don't get why you need to use the `.size()` to count from 0, 1, 2, ... And this map is quite weird.

**Comment by axic** [comment]
True, this should be enough:
```
unsigned index = -1;
for (auto const& s: m_sources)
    indices[s.first] = ++index;
```

**Comment by pirapira** [comment]
Took a minute to figure out why this changes things.  Yes, many side-effects.

**Comment by chriseth** [comment]
I just checked our coverity run, it did not find the issue. :-( There might be similar issues in other places.

File: libsolidity/interface/CompilerStack.cpp [hunk 1]
```diff
@@ -406,8 +406,9 @@ vector<string> CompilerStack::sourceNames() const
 map<string, unsigned> CompilerStack::sourceIndices() const
 {
 	map<string, unsigned> indices;
+	unsigned index = 0;
 	for (auto const& s: m_sources)
-		indices[s.first] = indices.size();
```
Comment (chriseth) [inline libsolidity/interface/CompilerStack.cpp:410]: `indices.insert(make_pair(s.first, indices.size()))` would not have that issue, but it's probably more readable like this.

**Comment by axic** [comment]
> There might be similar issues in other places.

```
./ast/Types.cpp:                memberTypes.reserve(m_memberTypes.size());
```

Nevermind, too late to notice the `m_` prefix.

File: libsolidity/interface/CompilerStack.cpp (hunk 1 above)
Comment (chfast) [reply libsolidity/interface/CompilerStack.cpp:410]: `indices.emplace(s.first, indices.size())`, or even better `indices.emplace(s.first, index++)` as discussed offline.

**Comment by chriseth** [comment]
I wonder if it would be more consistent to first sort them and then assign the indices. We could use some boost range adaptor magic for that (but we don't have to):
```
map<string, unsigned> indices;
for (auto i: m_sources | sorted | indexed(0))
    indices[i.value()] = i.index();
return indices;
```

**Comment by axic** [comment]
In that case `sourceNames()` must be sorted too.

**Comment by axic** [comment]
I think this should be merged as is, because sorting might have some other side-effects somewhere. I'm fine in having them sorted in a separate PR.

**Comment by chriseth** [comment]
`sourceNames` is sorted, iterating over a map returns the items sorted by key

**Comment by axic** [comment]
What does `m_sources | sorted | indexed(0)` sorted and indexed mean here?

**Comment by chriseth** [comment]
Ok but then my comment above is void. Just for some fun, something like the following could also work:
`return boost::copy_range<map<string, unsigned>>(m_sources | indexed(0))` :-)

**Comment by axic** [comment]
I would have no idea what `return boost::copy_range<map<string, unsigned>>(m_sources | indexed(0))` is doing 😉 

Looks fun though

**Pull request merged**